packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -1745,42 +2305,125 @@ class _SelectableFragment with Selectable, Diagnosticable, ChangeNotifier implem
     return SelectionResult.none;
   }
 
-  SelectionResult _handleSelectWord(Offset globalPosition) {
-    final TextPosition position = paragraph.getPositionForOffset(paragraph.globalToLocal(globalPosition));
-    if (_positionIsWithinCurrentSelection(position) && _textSelectionStart != _textSelectionEnd) {
-      return SelectionResult.end;
-    }
-    final _WordBoundaryRecord wordBoundary = _getWordBoundaryAtPosition(position);
-    // This fragment may not contain the word, decide what direction the target
+  SelectionResult _handleSelectTextBoundary(_TextBoundaryRecord textBoundary) {
+    // This fragment may not contain the boundary, decide what direction the target
     // fragment is located in. Because fragments are separated by placeholder
-    // spans, we also check if the beginning or end of the word is touching
+    // spans, we also check if the beginning or end of the boundary is touching
     // either edge of this fragment.
-    if (wordBoundary.wordStart.offset < range.start && wordBoundary.wordEnd.offset <= range.start) {
+    if (textBoundary.boundaryStart.offset < range.start && textBoundary.boundaryEnd.offset <= range.start) {
       return SelectionResult.previous;
-    } else if (wordBoundary.wordStart.offset >= range.end && wordBoundary.wordEnd.offset > range.end) {
+    } else if (textBoundary.boundaryStart.offset >= range.end && textBoundary.boundaryEnd.offset > range.end) {
       return SelectionResult.next;
     }
-    // Fragments are separated by placeholder span, the word boundary shouldn't
+    // Fragments are separated by placeholder span, the text boundary shouldn't
     // expand across fragments.
-    assert(wordBoundary.wordStart.offset >= range.start && wordBoundary.wordEnd.offset <= range.end);
-    _textSelectionStart = wordBoundary.wordStart;
-    _textSelectionEnd = wordBoundary.wordEnd;
-    _selectableContainsOriginWord = true;
+    assert(textBoundary.boundaryStart.offset >= range.start && textBoundary.boundaryEnd.offset <= range.end);
+    _textSelectionStart = textBoundary.boundaryStart;
+    _textSelectionEnd = textBoundary.boundaryEnd;
+    _selectableContainsOriginTextBoundary = true;
     return SelectionResult.end;
   }
 
-  _WordBoundaryRecord _getWordBoundaryAtPosition(TextPosition position) {
-    final TextRange word = paragraph.getWordBoundary(position);
-    assert(word.isNormalized);
-    late TextPosition start;
-    late TextPosition end;
-    if (position.offset > word.end) {
+  TextRange? _intersect(TextRange a, TextRange b) {
```

Comment:
Good point. Looking at the usage of this method, we always pass it normalized ranges. We could probably assert that both `a`, and `b` are normalized just in case.